### DISCUSSION TITLE
Calculation of Precision and Recall

### DISCUSSION THREAD

**Discussion by Aiki89** [ultralytics/ultralytics#13492] [Docs]
I am working with Ultralytics YOLOv8 for my Master's thesis. I could not find out how Precision and Recall are calculated from the documentation (e.g. which thresholds are used etc.). Do you have this information for me? Thank you in advance!

**Comment by pderrenger** [comment]
Hello!

Thank you for reaching out and for using Ultralytics YOLOv8 for your Master's thesis! Precision and Recall are indeed crucial metrics for evaluating the performance of object detection models, and I'm happy to provide some clarity on how they are calculated in YOLOv8.

### Precision and Recall Calculation

In YOLOv8, Precision and Recall are calculated based on the Intersection over Union (IoU) threshold, which determines whether a predicted bounding box is considered a true positive. Here’s a brief overview:

- **Precision** is the ratio of true positive detections to the total number of positive detections (true positives + false positives). It measures the accuracy of the positive predictions.
- **Recall** is the ratio of true positive detections to the total number of actual positives (true positives + false negatives). It measures the model's ability to detect all relevant instances.

### Thresholds Used

The default IoU threshold used for calculating these metrics in YOLOv8 is typically 0.5, which means a predicted bounding box is considered correct if it overlaps with the ground truth box by at least 50%. However, YOLOv8 also calculates metrics across a range of IoU thresholds (from 0.5 to 0.95 in steps of 0.05) to provide a more comprehensive evaluation, often referred to as mAP50-95.

### Example Code

Here’s a simple example of how you can validate your model and obtain these metrics using the `model.val()` function:

```python
from ultralytics import YOLO

# Load your trained model
model = YOLO('path/to/your/model.pt')

# Validate the model
metrics = model.val()

# Access precision and recall
precision = metrics.box.P
recall = metrics.box.R

print(f"Precision: {precision}")
print(f"Recall: {recall}")
```

### Further Reading

For a deeper dive into how these metrics are calculated and interpreted, I recommend checking out our [Performance Metrics Guide](https://docs.ultralytics.com/guides/yolo-performance-metrics/). This guide provides detailed explanations and visualizations that can help you better understand the evaluation process.

If you encounter any issues or need further assistance, please ensure you are using the latest versions of `torch` and `ultralytics`. If the problem persists, providing a minimum reproducible example will greatly help us in diagnosing and resolving the issue. You can find more information on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any more questions. Best of luck with your thesis! 😊
